Adds support for drawing 🎨 by integrating excalidraw

NikhilA8606 opened the pull request:
## Required Backends
- https://github.com/ohcnetwork/care/pull/2880

## Integrated Excalidraw for drawing Files

- Fixes #10616
- Added another tab called "Drawing" to draw and save the file here.

![Screenshot from 2025-03-06 00-06-59](https://github.com/user-attachments/assets/53dcf4ff-8149-4302-9fbe-b33b0c7e6495)

- This is the Excalidraw page interface

![Screenshot from 2025-03-06 00-09-23](https://github.com/user-attachments/assets/271c7a51-d451-4c39-ab36-b65e05ef5916)

After saving the drawing using filename users can view and update the file as they needed

![Screenshot from 2025-03-06 00-11-09](https://github.com/user-attachments/assets/3db97d56-519a-4db5-9dd1-db089ef8ade1)

This is the screenrecord of full process

[Screencast from 2025-03-06 00-15-01.webm](https://github.com/user-attachments/assets/90b444a8-9cf1-40e6-80cd-fc51f1593368)

@rithviknishad 

## Merge Checklist

- [x] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is kept in I18n files.
- [x] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [x] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced an intuitive drawing workflow for both patient and encounter sections with dedicated drawing tabs.
	- Added a drawing editor that lets users create, edit, and preview drawings, along with clear guidance and notifications on save/upload actions.
	- Enhanced the search and navigation capabilities to easily manage drawings.
	- Improved error handling and user notifications related to drawing actions.
	- Expanded the patient details interface with a dedicated section for drawings.
	- Added new routes for managing drawings in both patient and encounter contexts.
	- Integrated a new component for displaying a list of drawings associated with encounters or patients.
- **Bug Fixes**
	- Resolved issues related to file upload errors and drawing saving confirmations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

NikhilA8606 committed:
Integrated excalidraw component and add file save option for the page

NikhilA8606 committed:
Merge remote-tracking branch 'upstream/develop' into issues/#10616/Integrating-Excalidraw-for-drawing-files

NikhilA8606 committed:
Completed the File upload option and created new component for View the page

rithviknishad commented on src/Routers/routes/ConsultationRoutes.tsx:
```diff
@@ -6,6 +9,9 @@ import { AppRoutes } from "@/Routers/AppRouter";
 
+const ExcalidrawPage = lazy(() => import("@/components/Files/ExcalidrawPage"));
+const ExcalidrawView = lazy(() => import("@/components/Files/ExcalidrawView"));
+
```
Nice, but we could simplify things by lazy loading the `Excalidraw` component itself directly inside the `ExcalidrawPage` and `ExcalidrawView` component instead;

NikhilA8606 replied:
But what about the import("")); we have no Excalidraw.tsx (or .jsx) file inside the File

rithviknishad replied:
we could do something like this inside ExcalidrawEditor and Viewer since excalidraw doesn't have a default export

```tsx
const Excalidraw = lazy(() =>
  import("@excalidraw/excalidraw").then((module) => ({
    default: module.Excalidraw,
  })),
);
```

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+
+  const handleSave = async () => {
+    if (!name.trim()) {
+      toast.error("Please enter a name for the drawing.");
```
i18n

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+    const obj = {
+      type: "excalidraw",
+      version: "2",
+      source: "https://care.ohc.network.com",
```
1. `.network` is the TLD and not `.com`, this URL is incorrect
2. This should be extracted from the `window.location` since this would be deployed on domains other than `care.ohc.network`.

NikhilA8606 replied:
Yeah..got it

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+
+    try {
+      const file = new File([JSON.stringify(obj)], `${name}.excalidraw`, {
+        type: "text/plain",
```
ideally we should be using `application/vnd.excalidraw` or atleast `application/json`.

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+          initialData={{
+            appState: { theme: "light" },
+          }}
+          onChange={debounce(async (elements) => {
```
i don't see any reason why this needs to be `async`

```suggestion
          onChange={debounce((elements) => {
```

rithviknishad commented on src/Routers/routes/ConsultationRoutes.tsx:
```diff
@@ -6,6 +9,9 @@ import { AppRoutes } from "@/Routers/AppRouter";
 import { EncounterShow } from "@/pages/Encounters/EncounterShow";
 import { PrintPrescription } from "@/pages/Encounters/PrintPrescription";
 
+const ExcalidrawPage = lazy(() => import("@/components/Files/ExcalidrawPage"));
+const ExcalidrawView = lazy(() => import("@/components/Files/ExcalidrawView"));
```
namings could be confusing, keep it more explicit

```suggestion
const ExcalidrawEditor = lazy(() => import("@/components/Files/ExcalidrawPage"));
const ExcalidrawView = lazy(() => import("@/components/Files/ExcalidrawView"));
```

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+import mutate from "@/Utils/request/mutate";
+import query from "@/Utils/request/query";
+
+type Props = {
+  associatingId: string;
+  fileType: string;
+  drawingId?: string;
+};
```
`drawingId` is not used anywhere right? 

```suggestion
type Props = {
  associatingId: string;
  fileType: string;
};
```

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+  const [id, setId] = useState(drawingId);
+  const [isDirty, setIsDirty] = useState(false);
+
+  const { data, isLoading } = useQuery({
+    queryKey: ["file", id],
+    queryFn: query(routes.retrieveUpload, {
+      pathParams: { id: id || "" },
+    }),
+    enabled: !!id,
+  });
```
if `ExcalidrawPage` component can act as both viewer and editor, why do we need a separate component `ExcalidrawView`?

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+      });
+
+      if (!upload.ok) {
+        toast.error("Error uploading file");
```
missing i18n in many other places as well;

rithviknishad commented on src/components/Files/ExcalidrawPage.tsx:
```diff
@@ -0,0 +1,185 @@
+        navigate(`drawings/${response!.id}`);
+      }
+    } catch (error) {
+      console.error("Error in Step 1 (createUpload):", error);
```
```suggestion
```

rithviknishad commented on src/Routers/routes/ConsultationRoutes.tsx:
```diff
@@ -27,6 +33,20 @@ const consultationRoutes: AppRoutes = {
+      <ExcalidrawPage associatingId={encounterId} fileType="encounter" />
+    </Suspense>
+  ),
+  "/facility/:facilityId/patient/:patientId/encounter/:encounterId/drawings/:drawingId":
+    ({ drawingId }) => (
+      <Suspense fallback={<Loading />}>
+        <ExcalidrawView drawingId={drawingId} />
+      </Suspense>
+    ),
```
@bodhish marking this for hold since backend would not allow editing a file due to audit reasons.

@vigneshhari suggests making a plug and writing the elements to the DB itself instead of a file upload.

bodhish replied:
Buckets have versions what stops us from editing 🤔

NikhilA8606 committed:
Removed unncessary edit query in the Excalidraw editor page and other i18n conversions

NikhilA8606 committed:
Changed Frontend according to the Metartfact Model

rithviknishad committed:
fix loading focus issue

NikhilA8606 committed:
Added Excalidraw for patients also

NikhilA8606 committed:
Merge remote-tracking branch 'upstream/develop' into issues/#10616/Integrating-Excalidraw-for-drawing-files

NikhilA8606 committed:
Replced excalidraw with drawing and created Metaartifactrrquest for creat and update

nihal467 commented:
![image](https://github.com/user-attachments/assets/6798e21d-2b46-463b-8eaa-42234814fac0)

- [ ] when there is no drawing available its throwing 404 "Not Found" error in the frontend

rithviknishad committed:
Merge remote-tracking branch 'upstream/develop' into issues/#10616/Integrating-Excalidraw-for-drawing-files

rithviknishad committed:
cleanup code and other minor things

rithviknishad commented:
> ![image](https://private-user-images.githubusercontent.com/57055998/418864849-6798e21d-2b46-463b-8eaa-42234814fac0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExNTE1NDUsIm5iZiI6MTc0MTE1MTI0NSwicGF0aCI6Ii81NzA1NTk5OC80MTg4NjQ4NDktNjc5OGUyMWQtMmI0Ni00NjNiLThlYWEtNDIyMzQ4MTRmYWMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA1VDA1MDcyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlMDk3ZGMyMzA0ODg0MGE0MDUxMDY3MTBlOTEzZjVmY2Y4NTMxYTExYmIzOWY1YWJkZWY5MWNlNTE0MDNhNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.kXrKowPwWGb0w-aMoRbfboTL4tTPL8vj1M6gC3ZRGtg)
> 
> * [ ]  when there is no drawing available its throwing 404 "Not Found" error in the frontend
> 
> ![image](https://private-user-images.githubusercontent.com/57055998/418865503-822d5fcf-48c3-4b81-a76b-1317bce906cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExNTE1NDUsIm5iZiI6MTc0MTE1MTI0NSwicGF0aCI6Ii81NzA1NTk5OC80MTg4NjU1MDMtODIyZDVmY2YtNDhjMy00YjgxLWE3NmItMTMxN2JjZTkwNmNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA1VDA1MDcyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwMzk5ZGRiN2FmMDEwNWQ1MmE4YzE4ODUzMzE3NzlkNTg5ZDJkMjNiNjQ1YzNjMDQyYjJjZTQwY2ZhMTE2YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.twuv3N3v-J0zCZHkrcSqs1HpoCN0H8_A9G_cBHVTiSU)
> 
> * [ ]  meta artificate API is throwing 404, when we are trying to upload the file

Unable to reproduce this @nihal467

rithviknishad committed:
add text for no drawings

rithviknishad commented:
>in the view drawing page, grey out the file name and replace save button name with go back or cancel

let's not replace the button name, this might confuse user that this is auto save

rithviknishad committed:
fix responsiveness issue and udpate not working;

NikhilA8606 committed:
Add label to the saved file

rithviknishad committed:
fix unnecessary drawing id

rithviknishad committed:
feat: Add unsaved changes handling in Excalidraw editor

- Implement unsaved changes alert dialog with save/discard options
- Add back navigation with dirty state check
- Update translations for new dialog messages
- Improve editor UI layout and responsiveness
- Implement dynamic dirty state tracking using hashKey

rithviknishad committed:
feat: Enhance drawing list view with SVG preview and improved UI

- Add ExcalidrawPreview component for rendering drawing thumbnails
- Implement dynamic SVG generation with loading and empty states
- Improve drawing card layout with hover effects and clickable interactions
- Optimize grid layout for responsive design
- Add icons and improve metadata display for each drawing

nihal467 commented:
> ![image](https://private-user-images.githubusercontent.com/57055998/419784322-b256cd78-fd09-4aa6-a28e-89da28ad8d58.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEyNDM4MDQsIm5iZiI6MTc0MTI0MzUwNCwicGF0aCI6Ii81NzA1NTk5OC80MTk3ODQzMjItYjI1NmNkNzgtZmQwOS00YWE2LWEyOGUtODlkYTI4YWQ4ZDU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzA2VDA2NDUwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3NmU1ZTdkMmJlM2I0MWM4ODJiOWM4MTQ4OWEyOGM0NmViNTY0MjhlNjNjMTNmMjcxNTI2MjkwZTNmYjcwYzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.I4u2g3G-J4Qup490Kbo3aIuGkdZo7Yv-HbdsP8mwmPw)
> 
> @rithviknishad the excalidraw upload is not working,the upsert API is throwing 400 mentioning that the object cannot be viewed, check the screenshot

seem like a cache issue

nihal467 committed:
fixed the styling issue in the warning message

nihal467 commented:
![image](https://github.com/user-attachments/assets/c10729d9-65f1-4a22-ba87-12b60579d8a8)

- [ ] once the drawing is updated, the drawing preview only showing the updated view upon hard refresh (this can be taken as a separate PR)

nihal467 committed:
Merge branch 'develop' into issues/#10616/Integrating-Excalidraw-for-drawing-files

Jacobjeevan commented on src/components/Common/Drawings/DrawingTab.tsx:
```diff
@@ -0,0 +1,241 @@
+  facilityId: string;
+  patientId?: string;
+  encounter?: Encounter;
+  patient?: Patient;
```
why do we need this? 🤔 where's patient coming from anyways

Jacobjeevan commented on src/components/Common/Drawings/DrawingTab.tsx:
```diff
@@ -0,0 +1,241 @@
+          svgContainerRef.current.appendChild(svg);
+        }
+      } catch (error) {
+        console.error("Error generating SVG:", error);
```
toast instead

Jacobjeevan commented on src/Routers/routes/PatientRoutes.tsx:
```diff
@@ -44,6 +51,28 @@ const PatientRoutes: AppRoutes = {
+      </Suspense>
+    );
+  },
+  "/facility/:facilityId/patient/:patientId/drawings/:drawingId": ({
```
Add a route patient/id/drawings...as well (organization > patient view).

Jacobjeevan commented on src/components/Patient/PatientDetailsTab/index.tsx:
```diff
@@ -46,6 +47,10 @@ export const patientTabs = [
   },
+  {
+    route: "drawings",
+    component: DrawingTab,
```
Add a PatientDrawingsTab similar to Encounter one.

NikhilA8606 replied:
done

Jacobjeevan commented on src/components/Patient/PatientHome.tsx:
```diff
@@ -156,6 +156,7 @@ export const PatientHome = (props: {
           <div className="h-full lg:mr-7 lg:basis-5/6">
             {Tab && (
               <Tab
+                type="patient"
```
Remove/move into PatientDrawingsTab

NikhilA8606 committed:
Added PatientDrawingsTab

NikhilA8606 committed:
Resolved merge conflicts

NikhilA8606 committed:
Added correct url for encounter drawing page

NikhilA8606 committed:
Minor changes in the patient routing

rithviknishad commented on src/Routers/routes/PatientRoutes.tsx:
```diff
@@ -70,14 +71,8 @@ const PatientRoutes: AppRoutes = {
-      />
-    </Suspense>
+  "organization/:id/patient/:patientId/drawings": ({ patientId }) => (
+    <PatientDrawingTab patientId={patientId} />
```
is this right? 🤔

rithviknishad replied:
is this right? 🤔 

have you tested these out?

https://github.com/user-attachments/assets/7f587816-7b8f-4046-bb7b-00013c7d10cb

rithviknishad commented on src/components/Common/Drawings/DrawingTab.tsx:
```diff
@@ -0,0 +1,237 @@
+  );
+});
+
+// Add display name for debugging purposes
```
```suggestion
```

rithviknishad commented on src/components/Common/Drawings/DrawingTab.tsx:
```diff
@@ -0,0 +1,237 @@
+      isMounted = false;
+      clearTimeout(timeoutId);
+    };
+  }, [elements, svgKey]); // Add svgKey to dependencies
```
```suggestion
  }, [elements, svgKey]);
```

rithviknishad commented on src/components/Common/Drawings/DrawingTab.tsx:
```diff
@@ -0,0 +1,237 @@
+      }
+    };
+
+    // Small timeout to ensure DOM is ready, especially during hot reloading
```
```suggestion
```

rithviknishad commented on src/components/Common/Drawings/DrawingTab.tsx:
```diff
@@ -0,0 +1,237 @@
+                  key={drawing.id}
+                  className="overflow-hidden hover:shadow-md transition-shadow duration-200 group cursor-pointer"
+                  onClick={() => {
+                    if (props.type === "encounter") {
+                      navigate(
+                        `/facility/${props.encounter?.facility.id}/patient/${props.patientId}/encounter/${props.encounter?.id}/drawings/${drawing.id}`,
+                      );
+                    } else {
```
we no longer need an if check here right? this wouldn't work for org. use cases right?

NikhilA8606 committed:
Removed unwanted comments

NikhilA8606 committed:
Removed merge conflicts - removed and reinstall lock file

bodhish commented:
Should we move drwaings into the files tab? 🤔

NikhilA8606 commented:
> Should we move drawings into the files tab? 🤔
I’d like to do it as a separate PR

NikhilA8606 committed:
Resolved merge conflicts in the translational files

nihal467 commented:
LGTM

NikhilA8606 committed:
Resolved merge conflicts in package-lock.json file
